compiler/passes/src/const_propagation/const_propagate_expression.rs

```diff
@@ -0,0 +1,335 @@
+            if let Some(expr) = std::mem::take(&mut member.expression) {
+                member.expression = Some(self.reconstruct_expression(expr).0);
+            }
+        }
+        (Expression::Struct(input), Default::default())
+    }
+
+    fn reconstruct_ternary(&mut self, mut input: TernaryExpression) -> (Expression, Self::AdditionalOutput) {
+        let (cond, cond_value) = self.reconstruct_expression(*input.condition);
+
+        match cond_value {
+            Some(Value::Bool(true)) => self.reconstruct_expression(*input.if_true),
+            Some(Value::Bool(false)) => self.reconstruct_expression(*input.if_false),
+            _ => {
+                *input.condition = cond;
+                *input.if_true = self.reconstruct_expression(*input.if_true).0;
+                *input.if_false = self.reconstruct_expression(*input.if_false).0;
+                (Expression::Ternary(input), None)
+            }
+        }
+    }
+
+    fn reconstruct_access(&mut self, input: leo_ast::AccessExpression) -> (Expression, Self::AdditionalOutput) {
+        match input {
+            leo_ast::AccessExpression::Array(array) => self.reconstruct_array_access(array),
+            leo_ast::AccessExpression::AssociatedConstant(constant) => self.reconstruct_associated_constant(constant),
+            leo_ast::AccessExpression::AssociatedFunction(function) => self.reconstruct_associated_function(function),
+            leo_ast::AccessExpression::Member(member) => self.reconstruct_member_access(member),
+            leo_ast::AccessExpression::Tuple(tuple) => self.reconstruct_tuple_access(tuple),
+        }
+    }
+
+    fn reconstruct_array_access(&mut self, mut input: leo_ast::ArrayAccess) -> (Expression, Self::AdditionalOutput) {
+        let span = input.span();
+        *input.array = self.reconstruct_expression(*input.array).0;
+        let (index_expr, opt_value) = self.reconstruct_expression(*input.index);
+        if let Some(value) = opt_value {
+            // We can perform compile time bounds checking.
+
+            let ty = self.type_table.get(&input.array.id());
+            let Some(Type::Array(array_ty)) = ty else {
+                panic!("Type checking guaranteed that this is an array.");
+            };
+            let len = array_ty.length();
+
+            macro_rules! err {
+                ($x: expr) => {
+                    // Only emit a bounds error if we have no other errors yet.
+                    // This prevents a chain of redundant error messages when a loop is unrolled.
+                    if !self.handler.had_errors() {
+                        self.emit_err(StaticAnalyzerError::array_bounds($x, len, span));
+                    }
+                };
+            }
+
+            match value {
+                Value::U8(x) if x as usize >= len => err!(&x),
+                Value::U16(x) if x as usize >= len => err!(&x),
+                Value::U32(x) if x.try_into().unwrap_or(len) >= len => err!(&x),
```

Comment:
I think it would be okay to go up to `u32` as all array indices are required to be `u32`s.